Initial lib

mikev opened the pull request.

mikev committed:
create lorawan module

mikev committed:
parameterized unit tests

mikev committed:
create test folder

mikev committed:
update lorawan tests

mikev committed:
lorawan randomized tests

mikev committed:
minor refactor

mikev committed:
validate legacy Helium devaddr test

mikev committed:
minor refactor

mikev committed:
minor refactor

mikev committed:
simplify get_id

mikev committed:
add Rust translator

mikev committed:
export is_local_devaddr

mikev committed:
remove Rust code

mikev committed:
netid to parse_netid

mikev committed:
export everything

mikev committed:
add eunit test

mikev committed:
add public function comments

mikev committed:
checkpoint

mikev committed:
join_request

mikev committed:
join_accept

mikev committed:
fopts parsing

mikev committed:
remove dead code

mikev committed:
add lora_adr

mikev committed:
remove lorawan_vars

mikev committed:
encode_payload

mikev committed:
fport

mikev committed:
fctrlbits

mikev committed:
checkpoint

mikev committed:
payload_data

mikev committed:
checkpoint

mikev committed:
handle hex encoded strings

mikev committed:
checkpoint encode-decode cycle working

mikev committed:
add types

mikev committed:
minor cleanup

mikev committed:
checkpoint

mikev committed:
decode_encode

mikev committed:
join request

mikev committed:
join request working

mikev committed:
handle lowercase hex strings

mikev committed:
join accept payload_04_test

mikev committed:
payload_05_test

mikev committed:
fix warnings

mikev committed:
update crypto API calls

mikev committed:
checkpoint

mikev committed:
payload_06_test

mikev committed:
compute frames true fcnt

mikev committed:
exercise_test

mikev committed:
quiet tests

mikev committed:
minor test cleanup

mikev committed:
minor cleanup

mikev committed:
payload_all_test passes

mikev committed:
payload_08_test passes

mikev committed:
payload_09_test passes

mikev committed:
payload_11_test passes

mikev committed:
bespoke_fopts

mikev committed:
payload_13_test pass

mikev committed:
log_fopts

mikev committed:
issue_633_sequence

mikev committed:
add unit tests back to lorawan

mikev committed:
rename lorawan to lora_subnet

mikev committed:
rename lorawan module to lora_subnet

mikev committed:
rebar3 eunit PASS

mikev committed:
Merge branch 'develop' of https://github.com/helium/erlang-lorawan into initial-lib

mikev committed:
update crypto API

mikev committed:
fix cross reference warnings

mikev committed:
create lora_region

mikev committed:
create lora_region

mikev committed:
lorawan.hrl checkpoint

mikev committed:
compile checkpoint

mikev committed:
validate_channels

mikev committed:
exerise_plan checkpoint

mikev committed:
checkpoint eu868 and au915

mikev committed:
checkpoint eu868 and au915

mikev committed:
checkpoint IN865

mikev committed:
checkpoint CN470

mikev committed:
checkpoint KR920

mikev committed:
validate_tx_power

mikev committed:
sync with router

mikev committed:
rx2_datarate

mikev committed:
update rx2_datarate

mikev committed:
rebar3 xref OK

mikev committed:
rebar3 fmt checkpoint

mikev committed:
rebar3 fmt checkpoint

mikev committed:
rebar3 eunit checkpoint

mikev committed:
rebar3 fmt checkpoint

mikev committed:
rebar3 dialyzer checkpoint

mikev committed:
rebar3 dialyzer checkpoint

mikev committed:
rebar3 dialyzer checkpoint

mikev committed:
test eunit PASS

mikev committed:
checkpoint

mikev committed:
make test PASS

mikev committed:
rename headers

mikev committed:
rename to lora_plan

mikev committed:
data rate atoms

mikev committed:
datarate_to_index

mikev committed:
downlink_eirp

mikev committed:
max_uplink_snr

mikev committed:
tx_power_table

mikev committed:
rx_window

mikev committed:
window functions

mikev committed:
dr_to_down

mikev committed:
fix compile error

mikev committed:
fix merge conflict

mikev committed:
update max_payload_size

mikev committed:
max_uplink_payload_size and max_downlink_payload_size

mikev committed:
validate_payload_size eunit test

mikev committed:
eunit checkpoint

dpezely commented on src/lora_adr.erl:
```diff
@@ -0,0 +1,1425 @@
+%% code. Because of this, you have to bring your own IO (BYOIO). It
+%% will not perform any packet reception, transmission, or
+%% construction. Instead, you must call this module for every uplink
+%% packet received and the returned terms tell you want actions you
```
typo?

```suggestion
%% packet received and the returned terms tell you what actions you
```

mikev replied:
lora_adr is an **exact** copy of the original Router ADR engine code.  Before it was cloned into the LoRaWAN library I did add an extensive number of new eunit tests.  In the future we may rewrite or create a new module.

mikev committed:
rx2_window checkpoint

mikev committed:
join2_window checkpoint

dpezely commented on src/lora_adr.erl:
```diff
@@ -0,0 +1,1425 @@
+    Device;
+track_adr_answer(Device0, Answer) ->
+    #device{
+        %% TODO: how to handle the case where `PendingTail =/= []'?
```
> %% TODO: how to handle the case where `PendingTail =/= []'?

Maybe have this function return a tuple instead of a single `handle()`, much like `Result` or `Option` in Rust.  That could indicate that there's more, and then let the calling scope decide whether to iterate upon PendingTail or not.

mikev committed:
rx1_window checkpoint

dpezely commented on src/lora_adr.erl:
```diff
@@ -0,0 +1,1425 @@
+        MinTxPowerDBm,
+        MaxTxPowerDBm
+    ),
+    %% This part is tricky. Most places deal with
```
Can you be more specific?  places --> regions???  Countries?  Device vendors?
```suggestion
    %% This part is tricky. Most regions deal with
```

mikev committed:
join1_window checkpoint

dpezely commented on src/lora_plan.erl:
```diff
@@ -0,0 +1,848 @@
+-spec datarate_to_atom(binary()) -> atom().
+datarate_to_atom(Binary) ->
+    case Binary of
+        <<"SF12BW125">> -> 'SF12BW125';
```
Maybe you're already aware: there's `erlang:binary_to_existing_atom()`
which returns `badarg` upon no match that may be used for the catch-all `RFU`.  These atoms already exist elsewhere; e.g., line 432.

dpezely reviewed (commented):
This looks great!  
Thank you for use of intermediate vars for those of us less fluent on nuances of the LoRaWAN spec!

I realize that this is a Draft PR, but there's some code duplication; e.g., `lora_region.erl` starting L590 `tx_window()` vs `lora_plan.erl`, L264 with same comments, etc.
```erlang
-spec tx_window(atom(), #rxq{}, #txq{}, number()) -> #txq{}.
tx_window(?JOIN1_WINDOW, #rxq{tmms = Stamp}, TxQ, _RxDelaySeconds) when is_integer(Stamp) -> ...
```
vs
```erlang
-spec tx_window(atom(), #rxq{}, #txq{}, number()) -> #txq{}.
tx_window(?JOIN1_WINDOW, #rxq{tmms = Stamp}, TxQ, _RxDelaySeconds) when is_integer(Stamp) -> ...
```

mikev committed:
remove dead code

mikev committed:
test cleanup

mikev committed:
make test PASS

mikev committed:
export up_to_down_datarate

mikev committed:
make dialyzer OK

mikev committed:
bug for bug max_payload_size

mikev committed:
add AS923 variants

mikev committed:
create lora_chmask

mikev committed:
max_downlink_snr

mikev committed:
max_tx_power

mikev committed:
add missing payload sizes

mikev committed:
reorder plan definitions

mikev committed:
refactor datarate API

mikev committed:
refactor datarate API

mikev committed:
update max_uplink_snr

mikev committed:
rx1_offset min and max

mikev committed:
remove debug io:format

mikev committed:
add plan AS923_1B for Malaysia

mikev committed:
Update US915/AU915 to Sub-bank 2

mikev committed:
match uplink frequencies to packet_forwarder config

mikev committed:
test IN865 and KR920

michaeldjeffrey committed:
truncate floats to 4 places before checking for the downlink freq

When frequencies go to and from json they tend to get a bit mussed up.
We use 4 decimal places here for plans like IN865.

michaeldjeffrey committed:
Merge pull request #7 from helium/mj/up-freq-rounding-tolerance

Up frequency rounding tolerance

mikev committed:
Update IN865 and KR920 (#6)

* Update IN865 and KR920

* frequency rounding

* add eu433 plan

mikev committed:
Mv/fix us915 (#8)

* fix us915

* eunit exercise_snr

* Update src/lora_plan.erl

Co-authored-by: dpezely <dpezely@nova-labs.com>

Co-authored-by: dpezely <dpezely@nova-labs.com>

mikev committed:
LoRaWAN library improvements (#9)

* add fat channel tests

* unit test EU433 and CN470

* nearest function

* integrate nearest function

* test as923_2, as923_3 and as923_3

* minor test cleanup

* create build_link_adr_req

* add US915 and AU915 fat channels

* helium netid test

* little endian transform

* devaddr is little endian

* more little endian changes

* add devaddr_from_netid API

dpezely commented on test/lorawan_tests.erl:
```diff
@@ -0,0 +1,174 @@
+
+    NetID00 = 16#E00001,
+    NetID01 = 16#C00035,
+    %% Official NetID assigned to Helium by LoRa Alliance
```
It's not the Helium Foundation but the former Helium Systems, Inc.-- yes?
```suggestion
    %% Official NetID assigned to Nova Labs by LoRa Alliance
```

mikev replied:
lora_region is now renamed test_region.
Also test_region is now protected with EUNIIT ifdef

dpezely commented on src/lora_subnet.erl:
```diff
@@ -0,0 +1,650 @@
+    case Prefix band (1 bsl Index) of
+        0 -> 7 - Index;
+        _ -> netid_type(Prefix, Index - 1)
+    end.
+
+-spec nwk_addr(devaddr()) -> nwkaddr().
+nwk_addr(DevAddr) ->
+    NetID = the_netid(DevAddr),
+    AddrBitLen = addr_len(netid_class(NetID)),
+    IgnoreLen = 32 - AddrBitLen,
+    DevAddr2 = <<DevAddr:32/integer-unsigned>>,
+    <<_:IgnoreLen, NwkAddr:AddrBitLen/integer-unsigned>> = DevAddr2,
+    NwkAddr.
+
+-spec netid_addr_range(netid(), [netid()]) -> {non_neg_integer(), non_neg_integer()}.
+netid_addr_range(NetID, NetIDList0) ->
+    FoundNetID = lists:any(fun(X) -> X == NetID end, NetIDList0),
+    case FoundNetID of
+        false ->
+            {0, 0};
+        true ->
+            NetIDList = lists:takewhile(fun(X) -> X =/= NetID end, NetIDList0),
```
Combine similar list iterations: `lists:any()` before `case` and `lists:takewhile()` within the `true` branch can both be served by a single `takewhile()`.

```suggestion
    case lists:takewhile(fun(X) -> X =/= NetID end, NetIDList0) of
        [] ->
            {0, 0};
        NetIDList ->
```

mikev replied:
@dpezely , your change superficially seems reasonable, but it results in this test failure.  What you are missing is that  NetIDList can exist even when the NetID is not found...

  1) lora_subnet_tests:devaddr_test/0
     Failure/Error: ?assertEqual(1804050426, DevAddr2)
       expected: 1804050426
            got: 4286946170
     %% eunit_proc.erl:505:in `eunit_proc:handle_test/2`
     Output: DevAddr 6B879FFA
NetID 00000035
NetIDType 00000000
NetClass 00000000
AddrLen 00000019
IDLen 00000006
NwkAddr 01879FFA

     Output: DevAddr 6B879FFA
NetID 00000035
NetIDType 00000000
NetClass 00000000
AddrLen 00000019
IDLen 00000006
NwkAddr 01879FFA

dpezely commented on src/lora_region.erl:
```diff
@@ -0,0 +1,1627 @@
+%% Copyright (c) 2016-2019 Petr &lt;Gotthard petr.gotthard@@centrum.cz&gt;
+%% All rights reserved.
+%% Distributed under the terms of the MIT License. See the LICENSE file.
+%% @end
```
This copyright message no longer strictly covers this file's entire contents.

It's NOT simply "as-is" from the upstream source, considering there are a few comments mentioning `router` repo, etc. plus some familiar-looking code.

Maybe...?
```suggestion
%% Augmented for the Helium Network, 2022.
%% @end
```

mikev replied:
lora_region is an **exact** copy of the original router Region code.  It has several purposes but it should never be used outside this LoRaWAN library.  It is used for eunit testing only.
1. The lora_region functions are used to eunit test the new lora_plan functions.
2. It served to bootstrap the new lora_plan code because writing good test code is hard.
3. It is a useful sanity check on the new lora_plan functionality.

dpezely replied:
> lora_region is an exact copy of the original router Region code. It has several purposes but it should never be used outside this LoRaWAN library. It is used for eunit testing only.
>
> 1. The lora_region functions are used to eunit test the new lora_plan functions.
> 2. It served to bootstrap the new lora_plan code because writing good test code is hard.
> 3. It is a useful sanity check on the new lora_plan functionality.

Please add that text to top of the file.

mikev replied:
lora_region is now renamed test_region.
Also test_region is now protected with EUNIIT ifdef

dpezely commented on src/lora_region.erl:
```diff
@@ -0,0 +1,1627 @@
+    ?assertEqual(Ch, Ch2).
+
+valid_channel(Ch) ->
+    % valid_channel('EU868', Ch),
```
Why are there a few lines commented-out in this function?
If these are for illustrating non-applicable regions, please add a comment indicating this.  Otherwise, it appears as an experiment in progress... If that, please add `TODO` comment.

mikev replied:
Apologies, I should have mentioned.  There is no need to review lora_region.   Lora_region is only used internally for testing and it is an exact copy of the original Router Region code.

dpezely replied:
Since `lora_region.erl` is used only for regression testing, please move it from `src/` to `test/`.

That will eliminate false positives when trying to discover name of library functions when using `grep` or `rg` (ripgrep).

dpezely replied:
Following-up from Discord DM: 
Since there may be issues with stashing such a library under `test` subdirectory, maybe use a descriptive path like `src/regression-test-library_not-for-new-code/` instead, which should be obvious when using `grep` or `rg` (ripgrep).

dpezely replied:
Better yet, use a combination of **both subdirectory and module name** indicating that it's intended only for regression testing.

mikev replied:
lora_region is now renamed test_region.
Also test_region is now protected with EUNIIT ifdef

dpezely commented on src/lora_region.erl:
```diff
@@ -0,0 +1,1627 @@
+    (A =< Max) and (B >= Min).
+
+-endif.
+%% end of file
```
There are no tests exercising Rx Delay.

Some may be extracted from router repo: `test/router_SUITE.erl` such as:
-  rx_delay_join_test/1,
-  rx_delay_downlink_default_test/1,
-  rx_delay_change_ignored_by_device_downlink_test/1,
-  rx_delay_accepted_by_device_downlink_test/1,
-  rx_delay_continue_session_test/1,
-  rx_delay_change_during_session_test/1

mikev replied:
lora_region is now renamed test_region.
Also test_region is now protected with EUNIIT ifdef

dpezely reviewed (commented):
Both `lora_plan.erl` and `lora_region.erl` export functions with identical names but slightly different implementations:

-    join1_window/3,
-    join2_window/2,
-    rx1_window/4,
-    rx2_window/3,
-    rx1_or_rx2_window/4

If this is intentional, please add a comment near each of these (in both files) explaining significance of the implementation with respect to its alternate.

mikev committed:
Dual Plan (#13)

* Code to support the DualPlan, i.e. AS923_1 / AU915

mikev committed:
adjust EU868 max_eirp down to 16 (#14)

mikev committed:
refactor lora_plan eunit tests

mikev committed:
rename lora_region to test_region

mikev committed:
complete lora_region to test_region rename

mikev committed:
create lora_core_tests

mikev committed:
refactor lora_core eunit

mikev committed:
eunit test refactor

mikev committed:
refactor eunit lora_subnet_tests

mikev committed:
remove dead code

mikev committed:
AU915_DP and AU915_SB5

mikev committed:
Fold loops into one

Co-authored-by: dpezely <dpezely@nova-labs.com>

mikev committed:
Revert "Fold loops into one"

This reverts commit 5b9adc4ef939039a2d3bbdc4c26274e2be3311ea.

mikev committed:
merge from master

dpezely reviewed (approved):
Nice work!  That was a huge effort!
